### DISCUSSION TITLE
How will rational number be implemented?

### DISCUSSION THREAD

**Discussion by amano-kenji** [mattwparas/steel#151] [Q&A] [answered]
Will it be implemented as haskell's Decimal library or Raku's Rat type or something else?

**Comment by wmedrano** [comment, marked as answer]
Update: Related PR #155 

`Steel` newbie here but under the hood, it will probably be a combination [`Rational32`](https://docs.rs/num/latest/num/type.Rational32.html) and [`BigRational`](https://docs.rs/num/latest/num/type.BigRational.html). As for behavior:

```lisp
(+ 1/5 4/5) ;; 1
(- 1 1/5)   ;; 4/5
(/ 4 5)     ;; 4/5
(* 2/3 5/7) ;; 10/21
(+ 1/5 0.8) ;; 1.0
```

**Reply by amano-kenji** [reply]
BigRational can definitely do financial calculations although it seems less efficient than https://hackage.haskell.org/package/Decimal which uses big integer as an integer mantissa and an 8-bit unsigned integer as negative exponent.

Haskell Decimal library says

> Decimal numbers are represented as m*10^(-e) where m and e are integers. The exponent e is an unsigned Word8. Hence the smallest value that can be represented is 10^-255.

m is arbitrary precision integer.

The combination of big integer mantissa and (big int or limited) negative exponent seems more efficient and more rational than rational...

Maybe, I'm wrong. Anyhow, BigRational can still do the job.

Can I round a BigRational number to a fixed number of decimal places after arithmetic operations? For example, bitcoin has 8 decimal places. The US dollar has 2 decimal places.

**Reply by amano-kenji** [reply]
https://hackage.haskell.org/package/Decimal-0.5.2/docs/Data-Decimal.html#v:roundTo says

> Round a DecimalRaw to a specified number of decimal places.

This is what I need. I want to calculate the average of rational numbers and round it to a number of decimal places.

**Reply by wmedrano** [reply]
- The combination of big integer mantissa and (big int or limited) negative exponent seems more efficient and more rational than rational...

I'm not really making a judgement call here but fractions are how `Scheme` languages usually do things. We could also implement fractions under the hood to align more with the `Haskell` decimal package, but the [`num`](The combination of big integer mantissa and (big int or limited) negative exponent seems more efficient and more rational than rational...) crate already has `Rational` support so its just easier at the moment.

- Can I round a BigRational number ....

Not the greatest, but you can round to 8 decimal places with something like:

```lisp
(define (round-8 x) ;; TODO: Accept number of decimal places as an argument
  (let ((factor 100000000))
    (* factor (round (/ x factor)))))

(define (bitcoin-avg x y)
  (round-8 (/ (+ x y) 2))
```

For performance... maybe? It's hard to say with an interpreted language what the overhead will be. However, if we have some real life examples we can examine where Steel spends most of its time.

Note: That I think the performance will probably depend on if `FractV` or `BigFractV` is used. `FractV` should be a lot faster but requires that both the numerator and denominator are between `-2,147,483,648 and 2,147,483,647`.

- Rationals implementation

I have #155 in progress. If you have some example cases you wish to evaluate for performance, lmk.

**Reply by amano-kenji** [reply]
I don't know what exactly `round` does because I haven't used steel scheme, but `round-8` is useless to me. Different currencies use different numbers of decimal places. I want `round` for numbers with decimal places.

**Reply by wmedrano** [reply]
* I don't know what exactly `round` does:

- https://docs.racket-lang.org/reference/generic-numbers.html#%28def._%28%28quote._~23~25kernel%29._round%29%29

Round, rounds the number to the nearest whole number. For floating point numbers, it returns another float. For integers/fractions, it returns a fraction.

```lisp
(round 0.5) => 1.0
(round 1/2) => 1
(round 1/3) => 0
(round 10)  => 10
```

* `round-8` is useless to me
```lisp
(define (round-to-decimal x decimal-places)
  (let ((factor (expt 10 decimal-places)))
    (/ (round (* x factor))
       factor)))
```

**Reply by amano-kenji** [reply]
`round-to-decimal` seems to do what I want, but I am not sure that it is a good enough implementation for the core.

**Reply by mattwparas** [reply]
Why wouldn't it be a good enough implementation?

**Reply by amano-kenji** [reply]
That seems like a first draft. I know an optimized implementation when I see it.

**Reply by amano-kenji** [reply]
Consider adding

* big decimal which has big int mantissa and big int negative exponent
* decimal which has big int mantissa and 8-bit unsigned int negative exponent

Since big rational is already like big decimal, I guess adding decimal makes more sense. The purpose of decimal is efficiency for decimal numbers for up to 255 decimal places which should be enough for finance and enough in most cases.

Big rational and big decimal can get too big after division.

Perhaps, big rational division is more controllable than big decimal division?

**Reply by mattwparas** [reply]
> That seems like a first draft. I know an optimized implementation when I see it.

Care to suggest a more optimized implementation? I don't particularly see how we'd get something faster - it is pretty straightforward.

Why should these not be implemented in a library? As far as I'm aware, the numeric tower in scheme should be sufficient for any library to implement what you're describing.

**Reply by amano-kenji** [reply]
For something like haskell, a library is okay due to typeclasses. A math library can be used with core functions because core functions work with typeclasses, not specific types.

Without something like typeclasses, libraries don't really mix with core functions in my experiences....

For example, janet's persistent data structure library cannot be used with janet core functions.

**Reply by mattwparas** [reply]
Are you referring to math functions like `+`? Is the concern that `+` now won't work with your user defined math type?

**Reply by amano-kenji** [reply]
Anyway, if big rational can be prevented from growing too big upon division, decimal may not be important.

**Reply by amano-kenji** [reply]
Things like `+`, `-`, `*`, `/`, and `round-to-decimal`.

**Reply by mattwparas** [reply]
There is nothing stopping you (from the implementation stand point ) of implementing your own versions of these functions, and even shadow the existing definition within a module if you so choose.

**Reply by amano-kenji** [reply]
That's what I was worried about. So, there is no typeclass.

I guess I will have to manually economize memory consumption of big rational numbers. Things like eliminating common factors and rounding to a specific number of decimal places...

I can also implement my own operations in my library, but I prefer using core functions.

**Reply by wmedrano** [reply]
The implementation of rational numbers already eliminates common factors and even demotes from `BigRational` to `Rational32` when appropriate.

- [common factor removal](https://docs.rs/num/latest/num/rational/struct.Ratio.html#method.reduced)
- [`BigRational` to `Rational32` demotion](https://github.com/mattwparas/steel/blob/e779eb29e43b7bdd91934096b2f468679578bedc/crates/steel-core/src/primitives.rs#L279)

**Reply by amano-kenji** [reply]
You only need to implement something like `round-to-decimal` or `round-to`, then?

**Comment by kenaryn** [comment]
I apologize in advance if I am off-topic but would **valids** and **posits** (formerly known as **unum**)  be worth implementing? It is said to be more performant and accurate than floating point numbers and integers based on IEEE-754 specification.

- Here is the book describing the author's work: https://www.routledge.com/The-End-of-Error-Unum-Computing/Gustafson/p/book/9781482239867.
A glance at https://annas-archive.org/md5/b0e09ff07a6968086bcb3227ea08f95c shall make the trick.

- Here is the Rust port of posit numbers: https://crates.io/crates/softposit

- There are some Racket bindings for the C implementation: 
https://pkgs.racket-lang.org/package/softposit-rkt which I hope would be of value to help understand its internals and allow any motivated contributor to port it to Steel.
The associated documentation is there: https://pkgs.racket-lang.org/package/softposit-rkt/index.html

- Here is the C official implementation: https://gitlab.com/cerlane/SoftPosit published for NGA (New Generation Arithmetic collective).

- Here are some very brief lessons related to projective reals (from which the posits are a representation): https://web.archive.org/web/20170213233747/http://www.math.poly.edu/courses/projective_geometry/chapter_two/node1.html

And last but not least, the posit standard from march 2022: https://posithub.org/khub_doc

**Reply by mattwparas** [reply]
I see the value here, however given that steel is designed for first class embedding, I think straying from the builtin types would be difficult.

There shouldn't be any reason this couldn't be implemented as a library though! That being said if it were implemented separately as a shared library, we would probably lose all of the performance wins

**Reply by amano-kenji** [reply]
Will steel scheme embedded in helix editor be able to use steel scheme packages available for standalone interpreter?

I think there will be two kinds of packages.

* scheme steel packages for helix editor
* scheme steel packages for standalone interpreter

**Reply by mattwparas** [reply]
There shouldn't be any difference between the two. You should be able to use packages written for the standalone interpreter for  helix. Of course, if you write a package and it depends on APIs that only exist within helix, then it won't be able to be used by the standalone interpreter.
